### DISCUSSION TITLE
Diagnostic / Code Fixer idea

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#69966] [General]
I've noticed that in performance sensitive situations, dictionaries are often refactored into static switch statements.

It would be nice if there was a diagnostic/code fixer that could detect dictionaries that are never mutated and extract them out into a static switch-expression-based method.

**Comment by huoyaoyuan** [comment]
> I've noticed that in performance sensitive situations, dictionaries are often refactored into static switch statements.

It heavily depends on what the key/data type is. In fact for larger types, **switch statements should be refactored in to dictionary**.

Additionally, neither switch statement nor dictionary are performant enough for critical scenarios. A `smallint->bool` match may be manually optimized into a shift on magic constant.
